<div data-notebook-cell-style-version="2026-04-v2" style="color:#d7e0e7;line-height:1.8;font-family:ui-sans-serif,system-ui,-apple-system,BlinkMacSystemFont,'Segoe UI',sans-serif;">
<h1 style="background:linear-gradient(135deg,#22313d 0%,#284740 100%);color:#edf5f3;padding:18px 22px;border-radius:20px;border:1px solid #3a5255;border-left:10px solid #78b0a1;box-shadow:0 10px 24px rgba(0,0,0,0.20);margin:0 0 18px 0;">Pandas Cheat Sheet Notebook</h1>
<h2 style="background:linear-gradient(90deg,#22313d 0%,#273743 100%);border-left:8px solid #72aa9b;color:#e7eff3;padding:11px 16px;border-radius:14px;border:1px solid #3a4b58;box-shadow:0 6px 18px rgba(0,0,0,0.18);margin:22px 0 14px 0;">Advanced <code style="background:#27313a;color:#e8eef4;padding:2px 6px;border-radius:6px;border:1px solid #3c4a55;">groupby</code> explained from the start</h2>
<p style="margin:10px 0 15px 0;line-height:1.82;color:#d7e0e7;font-size:1rem;">This notebook is a clean interview-prep cheat sheet using <strong>one dataframe</strong> and solving <strong>Tasks 1 to 7</strong> with:</p>
<ul style="margin:10px 0 16px 20px;padding-left:18px;line-height:1.9;color:#d2dce4;">
<li style="margin:6px 0;">explanation</li>
<li style="margin:6px 0;">code</li>
<li style="margin:6px 0;">visible output</li>
<li style="margin:6px 0;">interview meaning</li>
</ul>
<p style="margin:10px 0 15px 0;line-height:1.82;color:#d7e0e7;font-size:1rem;">Topics covered:</p>
<ol style="margin:10px 0 16px 20px;padding-left:18px;line-height:1.9;color:#d2dce4;">
<li style="margin:6px 0;">Department-wise salary summary</li>
<li style="margin:6px 0;">Department and city wise average salary</li>
<li style="margin:6px 0;">Add department average salary to each row</li>
<li style="margin:6px 0;">Compare employee salary to department average</li>
<li style="margin:6px 0;">Count high performers by department</li>
<li style="margin:6px 0;">Percentage of high performers by department</li>
<li style="margin:6px 0;">Top paid employee in each department</li>
</ol>
</div>


<div data-notebook-cell-style-version="2026-04-v2" style="color:#d7e0e7;line-height:1.8;font-family:ui-sans-serif,system-ui,-apple-system,BlinkMacSystemFont,'Segoe UI',sans-serif;">
<h2 style="background:linear-gradient(90deg,#22313d 0%,#273743 100%);border-left:8px solid #72aa9b;color:#e7eff3;padding:11px 16px;border-radius:14px;border:1px solid #3a4b58;box-shadow:0 6px 18px rgba(0,0,0,0.18);margin:22px 0 14px 0;">0. Setup</h2>
<p style="margin:10px 0 15px 0;line-height:1.82;color:#d7e0e7;font-size:1rem;">We start by importing pandas and numpy.</p>
<ul style="margin:10px 0 16px 20px;padding-left:18px;line-height:1.9;color:#d2dce4;">
<li style="margin:6px 0;"><code style="background:#27313a;color:#e8eef4;padding:2px 6px;border-radius:6px;border:1px solid #3c4a55;">pandas</code> is for dataframe operations</li>
<li style="margin:6px 0;"><code style="background:#27313a;color:#e8eef4;padding:2px 6px;border-radius:6px;border:1px solid #3c4a55;">numpy</code> helps with conditional columns like <code style="background:#27313a;color:#e8eef4;padding:2px 6px;border-radius:6px;border:1px solid #3c4a55;">np.where</code></li>
</ul>
</div>


In [1]:
import pandas as pd
import numpy as np

<div data-notebook-cell-style-version="2026-04-v2" style="color:#d7e0e7;line-height:1.8;font-family:ui-sans-serif,system-ui,-apple-system,BlinkMacSystemFont,'Segoe UI',sans-serif;">
<h2 style="background:linear-gradient(90deg,#22313d 0%,#273743 100%);border-left:8px solid #72aa9b;color:#e7eff3;padding:11px 16px;border-radius:14px;border:1px solid #3a4b58;box-shadow:0 6px 18px rgba(0,0,0,0.18);margin:22px 0 14px 0;">1. Create the dataframe</h2>
<p style="margin:10px 0 15px 0;line-height:1.82;color:#d7e0e7;font-size:1rem;">This is the single dataframe used for all questions.</p>
<p style="margin:10px 0 15px 0;line-height:1.82;color:#d7e0e7;font-size:1rem;">Important columns:</p>
<ul style="margin:10px 0 16px 20px;padding-left:18px;line-height:1.9;color:#d2dce4;">
<li style="margin:6px 0;"><code style="background:#27313a;color:#e8eef4;padding:2px 6px;border-radius:6px;border:1px solid #3c4a55;">dept</code>: department</li>
<li style="margin:6px 0;"><code style="background:#27313a;color:#e8eef4;padding:2px 6px;border-radius:6px;border:1px solid #3c4a55;">city</code>: location</li>
<li style="margin:6px 0;"><code style="background:#27313a;color:#e8eef4;padding:2px 6px;border-radius:6px;border:1px solid #3c4a55;">salary</code>: fixed pay</li>
<li style="margin:6px 0;"><code style="background:#27313a;color:#e8eef4;padding:2px 6px;border-radius:6px;border:1px solid #3c4a55;">bonus</code>: extra pay</li>
<li style="margin:6px 0;"><code style="background:#27313a;color:#e8eef4;padding:2px 6px;border-radius:6px;border:1px solid #3c4a55;">rating</code>: performance rating</li>
<li style="margin:6px 0;"><code style="background:#27313a;color:#e8eef4;padding:2px 6px;border-radius:6px;border:1px solid #3c4a55;">skills</code>: list column</li>
<li style="margin:6px 0;"><code style="background:#27313a;color:#e8eef4;padding:2px 6px;border-radius:6px;border:1px solid #3c4a55;">projects</code>: nested list of dictionaries</li>
</ul>
<p style="margin:10px 0 15px 0;line-height:1.82;color:#d7e0e7;font-size:1rem;">We also convert <code style="background:#27313a;color:#e8eef4;padding:2px 6px;border-radius:6px;border:1px solid #3c4a55;">joined</code> into datetime.</p>
</div>


In [2]:
df = pd.DataFrame({
    "emp_id": [101, 102, 103, 104, 105, 106, 107, 108, 109, 110],
    "name": ["Alice", "Bob", "Charlie", "David", "Eva", "Frank", "Grace", "Helen", "Ian", "Jane"],
    "dept": ["Data", "Data", "Web", "Web", "Data", "ML", "ML", "Web", "Data", "ML"],
    "city": ["Bangalore", "Bangalore", "Delhi", "Delhi", "Mumbai", "Bangalore", "Mumbai", "Delhi", "Mumbai", "Bangalore"],
    "salary": [120000, 95000, 110000, 105000, 130000, 150000, 140000, 99000, 125000, 155000],
    "bonus": [10000, 8000, 12000, 7000, 15000, 20000, 18000, 6000, 14000, 22000],
    "rating": [5, 3, 4, 2, 5, 4, 5, 3, 4, 5],
    "experience": [6, 3, 5, 2, 7, 8, 6, 3, 5, 9],
    "joined": ["2020-01-15", "2021-06-10", "2019-09-23", "2022-03-01", "2018-11-12",
               "2017-07-19", "2020-08-25", "2021-01-11", "2019-05-30", "2016-12-05"],
    "skills": [
        ["python", "sql", "pandas"],
        ["python", "excel"],
        ["javascript", "react"],
        ["javascript", "css"],
        ["python", "sql", "spark"],
        ["python", "ml", "deep-learning"],
        ["python", "ml", "sql"],
        ["html", "css", "javascript"],
        ["python", "sql", "tableau"],
        ["python", "ml", "nlp"]
    ],
    "projects": [
        [{"name": "fraud", "hours": 120}, {"name": "churn", "hours": 80}],
        [{"name": "sales_dashboard", "hours": 60}],
        [{"name": "website_revamp", "hours": 100}],
        [{"name": "landing_page", "hours": 40}, {"name": "seo_fix", "hours": 30}],
        [{"name": "recommendation", "hours": 150}],
        [{"name": "forecasting", "hours": 200}, {"name": "vision", "hours": 90}],
        [{"name": "classification", "hours": 170}],
        [{"name": "frontend_migration", "hours": 110}],
        [{"name": "reporting", "hours": 95}, {"name": "automation", "hours": 50}],
        [{"name": "llm_search", "hours": 220}, {"name": "chatbot", "hours": 140}]
    ]
})

df["joined"] = pd.to_datetime(df["joined"])
df

,emp_id,name,dept,city,salary,bonus,rating,experience,joined,skills,projects
0,101,Alice,Data,Bangalore,120000,10000,5,6,2020-01-15,"[python, sql, pandas]","[{'name': 'fraud', 'hours': 120}, {'name': 'ch..."
1,102,Bob,Data,Bangalore,95000,8000,3,3,2021-06-10,"[python, excel]","[{'name': 'sales_dashboard', 'hours': 60}]"
2,103,Charlie,Web,Delhi,110000,12000,4,5,2019-09-23,"[javascript, react]","[{'name': 'website_revamp', 'hours': 100}]"
3,104,David,Web,Delhi,105000,7000,2,2,2022-03-01,"[javascript, css]","[{'name': 'landing_page', 'hours': 40}, {'name..."
4,105,Eva,Data,Mumbai,130000,15000,5,7,2018-11-12,"[python, sql, spark]","[{'name': 'recommendation', 'hours': 150}]"
5,106,Frank,ML,Bangalore,150000,20000,4,8,2017-07-19,"[python, ml, deep-learning]","[{'name': 'forecasting', 'hours': 200}, {'name..."
6,107,Grace,ML,Mumbai,140000,18000,5,6,2020-08-25,"[python, ml, sql]","[{'name': 'classification', 'hours': 170}]"
7,108,Helen,Web,Delhi,99000,6000,3,3,2021-01-11,"[html, css, javascript]","[{'name': 'frontend_migration', 'hours': 110}]"
8,109,Ian,Data,Mumbai,125000,14000,4,5,2019-05-30,"[python, sql, tableau]","[{'name': 'reporting', 'hours': 95}, {'name': ..."
9,110,Jane,ML,Bangalore,155000,22000,5,9,2016-12-05,"[python, ml, nlp]","[{'name': 'llm_search', 'hours': 220}, {'name'..."


<div data-notebook-cell-style-version="2026-04-v2" style="color:#d7e0e7;line-height:1.8;font-family:ui-sans-serif,system-ui,-apple-system,BlinkMacSystemFont,'Segoe UI',sans-serif;">
<h2 style="background:linear-gradient(90deg,#22313d 0%,#273743 100%);border-left:8px solid #72aa9b;color:#e7eff3;padding:11px 16px;border-radius:14px;border:1px solid #3a4b58;box-shadow:0 6px 18px rgba(0,0,0,0.18);margin:22px 0 14px 0;">Quick reminder: what <code style="background:#27313a;color:#e8eef4;padding:2px 6px;border-radius:6px;border:1px solid #3c4a55;">groupby</code> does</h2>
<p style="margin:10px 0 15px 0;line-height:1.82;color:#d7e0e7;font-size:1rem;"><code style="background:#27313a;color:#e8eef4;padding:2px 6px;border-radius:6px;border:1px solid #3c4a55;">groupby</code> means:</p>
<ol style="margin:10px 0 16px 20px;padding-left:18px;line-height:1.9;color:#d2dce4;">
<li style="margin:6px 0;">split rows into groups</li>
<li style="margin:6px 0;">apply a function to each group</li>
<li style="margin:6px 0;">combine the results</li>
</ol>
<h3 style="display:inline-block;color:#d8e4ea;background:#22303a;border-left:5px solid #7fa89e;padding:8px 14px;border-radius:12px;border:1px solid #3a4a52;margin:18px 0 10px 0;">Common patterns</h3>
<ul style="margin:10px 0 16px 20px;padding-left:18px;line-height:1.9;color:#d2dce4;">
<li style="margin:6px 0;"><code style="background:#27313a;color:#e8eef4;padding:2px 6px;border-radius:6px;border:1px solid #3c4a55;">agg()</code> -&gt; one row per group</li>
<li style="margin:6px 0;"><code style="background:#27313a;color:#e8eef4;padding:2px 6px;border-radius:6px;border:1px solid #3c4a55;">transform()</code> -&gt; same number of rows as original dataframe</li>
<li style="margin:6px 0;"><code style="background:#27313a;color:#e8eef4;padding:2px 6px;border-radius:6px;border:1px solid #3c4a55;">filter()</code> -&gt; keep or remove whole groups</li>
<li style="margin:6px 0;"><code style="background:#27313a;color:#e8eef4;padding:2px 6px;border-radius:6px;border:1px solid #3c4a55;">apply()</code> -&gt; custom logic on each group</li>
</ul>
</div>


<div data-notebook-cell-style-version="2026-04-v2" style="color:#d7e0e7;line-height:1.8;font-family:ui-sans-serif,system-ui,-apple-system,BlinkMacSystemFont,'Segoe UI',sans-serif;">
<h2 style="background:linear-gradient(90deg,#22313d 0%,#273743 100%);border-left:8px solid #72aa9b;color:#e7eff3;padding:11px 16px;border-radius:14px;border:1px solid #3a4b58;box-shadow:0 6px 18px rgba(0,0,0,0.18);margin:22px 0 14px 0;">2. Task 1 — Department-wise salary summary</h2>
</div>


<div data-notebook-cell-style-version="2026-04-v2" style="color:#d7e0e7;line-height:1.8;font-family:ui-sans-serif,system-ui,-apple-system,BlinkMacSystemFont,'Segoe UI',sans-serif;">
<h3 style="display:inline-block;color:#d8e4ea;background:#22303a;border-left:5px solid #7fa89e;padding:8px 14px;border-radius:12px;border:1px solid #3a4a52;margin:18px 0 10px 0;">Goal</h3>
<p style="margin:10px 0 15px 0;line-height:1.82;color:#d7e0e7;font-size:1rem;">For each <code style="background:#27313a;color:#e8eef4;padding:2px 6px;border-radius:6px;border:1px solid #3c4a55;">dept</code>, calculate:</p>
<ul style="margin:10px 0 16px 20px;padding-left:18px;line-height:1.9;color:#d2dce4;">
<li style="margin:6px 0;">total salary</li>
<li style="margin:6px 0;">average salary</li>
<li style="margin:6px 0;">max salary</li>
<li style="margin:6px 0;">min salary</li>
<li style="margin:6px 0;">employee count</li>
</ul>
<h3 style="display:inline-block;color:#d8e4ea;background:#22303a;border-left:5px solid #7fa89e;padding:8px 14px;border-radius:12px;border:1px solid #3a4a52;margin:18px 0 10px 0;">Why this uses <code style="background:#27313a;color:#e8eef4;padding:2px 6px;border-radius:6px;border:1px solid #3c4a55;">agg</code></h3>
<p style="margin:10px 0 15px 0;line-height:1.82;color:#d7e0e7;font-size:1rem;">We want <strong>one summary row per department</strong>, so <code style="background:#27313a;color:#e8eef4;padding:2px 6px;border-radius:6px;border:1px solid #3c4a55;">groupby(...).agg(...)</code> is the right tool.</p>
</div>


In [3]:
dept_salary_summary = df.groupby("dept").agg(
    total_salary=("salary", "sum"),
    avg_salary=("salary", "mean"),
    max_salary=("salary", "max"),
    min_salary=("salary", "min"),
    employee_count=("emp_id", "count")
).reset_index()

dept_salary_summary

,dept,total_salary,avg_salary,max_salary,min_salary,employee_count
0,Data,470000,117500.000000,130000,95000,4
1,ML,445000,148333.333333,155000,140000,3
2,Web,314000,104666.666667,110000,99000,3


<div data-notebook-cell-style-version="2026-04-v2" style="color:#d7e0e7;line-height:1.8;font-family:ui-sans-serif,system-ui,-apple-system,BlinkMacSystemFont,'Segoe UI',sans-serif;">
<h3 style="display:inline-block;color:#d8e4ea;background:#22303a;border-left:5px solid #7fa89e;padding:8px 14px;border-radius:12px;border:1px solid #3a4a52;margin:18px 0 10px 0;">Explanation</h3>
<p style="margin:10px 0 15px 0;line-height:1.82;color:#d7e0e7;font-size:1rem;">For <code style="background:#27313a;color:#e8eef4;padding:2px 6px;border-radius:6px;border:1px solid #3c4a55;">Data</code> department, salaries are:</p>
<ul style="margin:10px 0 16px 20px;padding-left:18px;line-height:1.9;color:#d2dce4;">
<li style="margin:6px 0;">120000</li>
<li style="margin:6px 0;">95000</li>
<li style="margin:6px 0;">130000</li>
<li style="margin:6px 0;">125000</li>
</ul>
<p style="margin:10px 0 15px 0;line-height:1.82;color:#d7e0e7;font-size:1rem;">So:</p>
<ul style="margin:10px 0 16px 20px;padding-left:18px;line-height:1.9;color:#d2dce4;">
<li style="margin:6px 0;">total = 470000</li>
<li style="margin:6px 0;">average = 117500</li>
<li style="margin:6px 0;">max = 130000</li>
<li style="margin:6px 0;">min = 95000</li>
<li style="margin:6px 0;">count = 4</li>
</ul>
<p style="margin:10px 0 15px 0;line-height:1.82;color:#d7e0e7;font-size:1rem;">This is a classic interview pattern: <strong>group + multiple metrics</strong>.</p>
</div>


<div data-notebook-cell-style-version="2026-04-v2" style="color:#d7e0e7;line-height:1.8;font-family:ui-sans-serif,system-ui,-apple-system,BlinkMacSystemFont,'Segoe UI',sans-serif;">
<h2 style="background:linear-gradient(90deg,#22313d 0%,#273743 100%);border-left:8px solid #72aa9b;color:#e7eff3;padding:11px 16px;border-radius:14px;border:1px solid #3a4b58;box-shadow:0 6px 18px rgba(0,0,0,0.18);margin:22px 0 14px 0;">3. Task 2 — Department and city wise average salary</h2>
</div>


<div data-notebook-cell-style-version="2026-04-v2" style="color:#d7e0e7;line-height:1.8;font-family:ui-sans-serif,system-ui,-apple-system,BlinkMacSystemFont,'Segoe UI',sans-serif;">
<h3 style="display:inline-block;color:#d8e4ea;background:#22303a;border-left:5px solid #7fa89e;padding:8px 14px;border-radius:12px;border:1px solid #3a4a52;margin:18px 0 10px 0;">Goal</h3>
<p style="margin:10px 0 15px 0;line-height:1.82;color:#d7e0e7;font-size:1rem;">Find average salary for each combination of:</p>
<ul style="margin:10px 0 16px 20px;padding-left:18px;line-height:1.9;color:#d2dce4;">
<li style="margin:6px 0;"><code style="background:#27313a;color:#e8eef4;padding:2px 6px;border-radius:6px;border:1px solid #3c4a55;">dept</code></li>
<li style="margin:6px 0;"><code style="background:#27313a;color:#e8eef4;padding:2px 6px;border-radius:6px;border:1px solid #3c4a55;">city</code></li>
</ul>
<h3 style="display:inline-block;color:#d8e4ea;background:#22303a;border-left:5px solid #7fa89e;padding:8px 14px;border-radius:12px;border:1px solid #3a4a52;margin:18px 0 10px 0;">Why this uses multi-column <code style="background:#27313a;color:#e8eef4;padding:2px 6px;border-radius:6px;border:1px solid #3c4a55;">groupby</code></h3>
<p style="margin:10px 0 15px 0;line-height:1.82;color:#d7e0e7;font-size:1rem;">Now the group is not just department.<br />
The group is the combination of <strong>department + city</strong>.</p>
</div>


In [4]:
dept_city_avg_salary = df.groupby(["dept", "city"], as_index=False).agg(
    avg_salary=("salary", "mean")
)

dept_city_avg_salary

,dept,city,avg_salary
0,Data,Bangalore,107500.000000
1,Data,Mumbai,127500.000000
2,ML,Bangalore,152500.000000
3,ML,Mumbai,140000.000000
4,Web,Delhi,104666.666667


<div data-notebook-cell-style-version="2026-04-v2" style="color:#d7e0e7;line-height:1.8;font-family:ui-sans-serif,system-ui,-apple-system,BlinkMacSystemFont,'Segoe UI',sans-serif;">
<h3 style="display:inline-block;color:#d8e4ea;background:#22303a;border-left:5px solid #7fa89e;padding:8px 14px;border-radius:12px;border:1px solid #3a4a52;margin:18px 0 10px 0;">Explanation</h3>
<p style="margin:10px 0 15px 0;line-height:1.82;color:#d7e0e7;font-size:1rem;">Example:</p>
<ul style="margin:10px 0 16px 20px;padding-left:18px;line-height:1.9;color:#d2dce4;">
<li style="margin:6px 0;"><code style="background:#27313a;color:#e8eef4;padding:2px 6px;border-radius:6px;border:1px solid #3c4a55;">Data + Bangalore</code> has Alice and Bob</li>
<li style="margin:6px 0;">average salary = <code style="background:#27313a;color:#e8eef4;padding:2px 6px;border-radius:6px;border:1px solid #3c4a55;">(120000 + 95000) / 2 = 107500</code></li>
</ul>
<p style="margin:10px 0 15px 0;line-height:1.82;color:#d7e0e7;font-size:1rem;">This pattern is useful when interviewers ask:</p>
<ul style="margin:10px 0 16px 20px;padding-left:18px;line-height:1.9;color:#d2dce4;">
<li style="margin:6px 0;">average by team and region</li>
<li style="margin:6px 0;">sales by category and month</li>
<li style="margin:6px 0;">score by class and section</li>
</ul>
</div>


<div data-notebook-cell-style-version="2026-04-v2" style="color:#d7e0e7;line-height:1.8;font-family:ui-sans-serif,system-ui,-apple-system,BlinkMacSystemFont,'Segoe UI',sans-serif;">
<h2 style="background:linear-gradient(90deg,#22313d 0%,#273743 100%);border-left:8px solid #72aa9b;color:#e7eff3;padding:11px 16px;border-radius:14px;border:1px solid #3a4b58;box-shadow:0 6px 18px rgba(0,0,0,0.18);margin:22px 0 14px 0;">4. Task 3 — Add department average salary to each row</h2>
</div>


<div data-notebook-cell-style-version="2026-04-v2" style="color:#d7e0e7;line-height:1.8;font-family:ui-sans-serif,system-ui,-apple-system,BlinkMacSystemFont,'Segoe UI',sans-serif;">
<h3 style="display:inline-block;color:#d8e4ea;background:#22303a;border-left:5px solid #7fa89e;padding:8px 14px;border-radius:12px;border:1px solid #3a4a52;margin:18px 0 10px 0;">Goal</h3>
<p style="margin:10px 0 15px 0;line-height:1.82;color:#d7e0e7;font-size:1rem;">Create a new column:</p>
<ul style="margin:10px 0 16px 20px;padding-left:18px;line-height:1.9;color:#d2dce4;">
<li style="margin:6px 0;"><code style="background:#27313a;color:#e8eef4;padding:2px 6px;border-radius:6px;border:1px solid #3c4a55;">dept_avg_salary</code></li>
</ul>
<h3 style="display:inline-block;color:#d8e4ea;background:#22303a;border-left:5px solid #7fa89e;padding:8px 14px;border-radius:12px;border:1px solid #3a4a52;margin:18px 0 10px 0;">Why this uses <code style="background:#27313a;color:#e8eef4;padding:2px 6px;border-radius:6px;border:1px solid #3c4a55;">transform</code></h3>
<p style="margin:10px 0 15px 0;line-height:1.82;color:#d7e0e7;font-size:1rem;">We want the <strong>department average repeated back onto every original row</strong>.</p>
<p style="margin:10px 0 15px 0;line-height:1.82;color:#d7e0e7;font-size:1rem;">That means:</p>
<ul style="margin:10px 0 16px 20px;padding-left:18px;line-height:1.9;color:#d2dce4;">
<li style="margin:6px 0;">same number of rows as original dataframe</li>
<li style="margin:6px 0;">so use <code style="background:#27313a;color:#e8eef4;padding:2px 6px;border-radius:6px;border:1px solid #3c4a55;">transform</code>, not <code style="background:#27313a;color:#e8eef4;padding:2px 6px;border-radius:6px;border:1px solid #3c4a55;">agg</code></li>
</ul>
</div>


In [5]:
df["dept_avg_salary"] = df.groupby("dept")["salary"].transform("mean")

df[["emp_id", "name", "dept", "salary", "dept_avg_salary"]]

,emp_id,name,dept,salary,dept_avg_salary
0,101,Alice,Data,120000,117500.000000
1,102,Bob,Data,95000,117500.000000
2,103,Charlie,Web,110000,104666.666667
3,104,David,Web,105000,104666.666667
4,105,Eva,Data,130000,117500.000000
5,106,Frank,ML,150000,148333.333333
6,107,Grace,ML,140000,148333.333333
7,108,Helen,Web,99000,104666.666667
8,109,Ian,Data,125000,117500.000000
9,110,Jane,ML,155000,148333.333333


<div data-notebook-cell-style-version="2026-04-v2" style="color:#d7e0e7;line-height:1.8;font-family:ui-sans-serif,system-ui,-apple-system,BlinkMacSystemFont,'Segoe UI',sans-serif;">
<h3 style="display:inline-block;color:#d8e4ea;background:#22303a;border-left:5px solid #7fa89e;padding:8px 14px;border-radius:12px;border:1px solid #3a4a52;margin:18px 0 10px 0;">Explanation</h3>
<p style="margin:10px 0 15px 0;line-height:1.82;color:#d7e0e7;font-size:1rem;">For the <code style="background:#27313a;color:#e8eef4;padding:2px 6px;border-radius:6px;border:1px solid #3c4a55;">Data</code> department, the average salary is <code style="background:#27313a;color:#e8eef4;padding:2px 6px;border-radius:6px;border:1px solid #3c4a55;">117500</code>.<br />
So every <code style="background:#27313a;color:#e8eef4;padding:2px 6px;border-radius:6px;border:1px solid #3c4a55;">Data</code> employee gets <code style="background:#27313a;color:#e8eef4;padding:2px 6px;border-radius:6px;border:1px solid #3c4a55;">117500</code> in the new column.</p>
<p style="margin:10px 0 15px 0;line-height:1.82;color:#d7e0e7;font-size:1rem;">Memory trick:</p>
<ul style="margin:10px 0 16px 20px;padding-left:18px;line-height:1.9;color:#d2dce4;">
<li style="margin:6px 0;"><code style="background:#27313a;color:#e8eef4;padding:2px 6px;border-radius:6px;border:1px solid #3c4a55;">agg()</code> shrinks</li>
<li style="margin:6px 0;"><code style="background:#27313a;color:#e8eef4;padding:2px 6px;border-radius:6px;border:1px solid #3c4a55;">transform()</code> broadcasts back</li>
</ul>
</div>


<div data-notebook-cell-style-version="2026-04-v2" style="color:#d7e0e7;line-height:1.8;font-family:ui-sans-serif,system-ui,-apple-system,BlinkMacSystemFont,'Segoe UI',sans-serif;">
<h2 style="background:linear-gradient(90deg,#22313d 0%,#273743 100%);border-left:8px solid #72aa9b;color:#e7eff3;padding:11px 16px;border-radius:14px;border:1px solid #3a4b58;box-shadow:0 6px 18px rgba(0,0,0,0.18);margin:22px 0 14px 0;">5. Task 4 — Compare employee salary to department average</h2>
</div>


<div data-notebook-cell-style-version="2026-04-v2" style="color:#d7e0e7;line-height:1.8;font-family:ui-sans-serif,system-ui,-apple-system,BlinkMacSystemFont,'Segoe UI',sans-serif;">
<h3 style="display:inline-block;color:#d8e4ea;background:#22303a;border-left:5px solid #7fa89e;padding:8px 14px;border-radius:12px;border:1px solid #3a4a52;margin:18px 0 10px 0;">Goal</h3>
<p style="margin:10px 0 15px 0;line-height:1.82;color:#d7e0e7;font-size:1rem;">Create a new column:</p>
<ul style="margin:10px 0 16px 20px;padding-left:18px;line-height:1.9;color:#d2dce4;">
<li style="margin:6px 0;"><code style="background:#27313a;color:#e8eef4;padding:2px 6px;border-radius:6px;border:1px solid #3c4a55;">salary_vs_dept_avg</code></li>
</ul>
<p style="margin:10px 0 15px 0;line-height:1.82;color:#d7e0e7;font-size:1rem;">Logic:</p>
<ul style="margin:10px 0 16px 20px;padding-left:18px;line-height:1.9;color:#d2dce4;">
<li style="margin:6px 0;"><code style="background:#27313a;color:#e8eef4;padding:2px 6px;border-radius:6px;border:1px solid #3c4a55;">&quot;above&quot;</code> if employee salary &gt; department average</li>
<li style="margin:6px 0;"><code style="background:#27313a;color:#e8eef4;padding:2px 6px;border-radius:6px;border:1px solid #3c4a55;">&quot;below_or_equal&quot;</code> otherwise</li>
</ul>
<h3 style="display:inline-block;color:#d8e4ea;background:#22303a;border-left:5px solid #7fa89e;padding:8px 14px;border-radius:12px;border:1px solid #3a4a52;margin:18px 0 10px 0;">Why this uses <code style="background:#27313a;color:#e8eef4;padding:2px 6px;border-radius:6px;border:1px solid #3c4a55;">np.where</code></h3>
<p style="margin:10px 0 15px 0;line-height:1.82;color:#d7e0e7;font-size:1rem;"><code style="background:#27313a;color:#e8eef4;padding:2px 6px;border-radius:6px;border:1px solid #3c4a55;">np.where(condition, value_if_true, value_if_false)</code> is a fast way to build conditional columns.</p>
</div>


In [6]:
df["salary_vs_dept_avg"] = np.where(
    df["salary"] > df["dept_avg_salary"],
    "above",
    "below_or_equal"
)

df[["emp_id", "name", "dept", "salary", "dept_avg_salary", "salary_vs_dept_avg"]]

,emp_id,name,dept,salary,dept_avg_salary,salary_vs_dept_avg
0,101,Alice,Data,120000,117500.000000,above
1,102,Bob,Data,95000,117500.000000,below_or_equal
2,103,Charlie,Web,110000,104666.666667,above
3,104,David,Web,105000,104666.666667,above
4,105,Eva,Data,130000,117500.000000,above
5,106,Frank,ML,150000,148333.333333,above
6,107,Grace,ML,140000,148333.333333,below_or_equal
7,108,Helen,Web,99000,104666.666667,below_or_equal
8,109,Ian,Data,125000,117500.000000,above
9,110,Jane,ML,155000,148333.333333,above


<div data-notebook-cell-style-version="2026-04-v2" style="color:#d7e0e7;line-height:1.8;font-family:ui-sans-serif,system-ui,-apple-system,BlinkMacSystemFont,'Segoe UI',sans-serif;">
<h3 style="display:inline-block;color:#d8e4ea;background:#22303a;border-left:5px solid #7fa89e;padding:8px 14px;border-radius:12px;border:1px solid #3a4a52;margin:18px 0 10px 0;">Explanation</h3>
<p style="margin:10px 0 15px 0;line-height:1.82;color:#d7e0e7;font-size:1rem;">For Bob in <code style="background:#27313a;color:#e8eef4;padding:2px 6px;border-radius:6px;border:1px solid #3c4a55;">Data</code>:</p>
<ul style="margin:10px 0 16px 20px;padding-left:18px;line-height:1.9;color:#d2dce4;">
<li style="margin:6px 0;">salary = 95000</li>
<li style="margin:6px 0;">department average = 117500</li>
<li style="margin:6px 0;">result = <code style="background:#27313a;color:#e8eef4;padding:2px 6px;border-radius:6px;border:1px solid #3c4a55;">below_or_equal</code></li>
</ul>
<p style="margin:10px 0 15px 0;line-height:1.82;color:#d7e0e7;font-size:1rem;">For Eva in <code style="background:#27313a;color:#e8eef4;padding:2px 6px;border-radius:6px;border:1px solid #3c4a55;">Data</code>:</p>
<ul style="margin:10px 0 16px 20px;padding-left:18px;line-height:1.9;color:#d2dce4;">
<li style="margin:6px 0;">salary = 130000</li>
<li style="margin:6px 0;">department average = 117500</li>
<li style="margin:6px 0;">result = <code style="background:#27313a;color:#e8eef4;padding:2px 6px;border-radius:6px;border:1px solid #3c4a55;">above</code></li>
</ul>
</div>


<div data-notebook-cell-style-version="2026-04-v2" style="color:#d7e0e7;line-height:1.8;font-family:ui-sans-serif,system-ui,-apple-system,BlinkMacSystemFont,'Segoe UI',sans-serif;">
<h2 style="background:linear-gradient(90deg,#22313d 0%,#273743 100%);border-left:8px solid #72aa9b;color:#e7eff3;padding:11px 16px;border-radius:14px;border:1px solid #3a4b58;box-shadow:0 6px 18px rgba(0,0,0,0.18);margin:22px 0 14px 0;">6. Task 5 — Count high performers by department</h2>
</div>


<div data-notebook-cell-style-version="2026-04-v2" style="color:#d7e0e7;line-height:1.8;font-family:ui-sans-serif,system-ui,-apple-system,BlinkMacSystemFont,'Segoe UI',sans-serif;">
<h3 style="display:inline-block;color:#d8e4ea;background:#22303a;border-left:5px solid #7fa89e;padding:8px 14px;border-radius:12px;border:1px solid #3a4a52;margin:18px 0 10px 0;">Goal</h3>
<p style="margin:10px 0 15px 0;line-height:1.82;color:#d7e0e7;font-size:1rem;">For each department, count employees where:</p>
<ul style="margin:10px 0 16px 20px;padding-left:18px;line-height:1.9;color:#d2dce4;">
<li style="margin:6px 0;"><code style="background:#27313a;color:#e8eef4;padding:2px 6px;border-radius:6px;border:1px solid #3c4a55;">rating &gt;= 4</code></li>
</ul>
<h3 style="display:inline-block;color:#d8e4ea;background:#22303a;border-left:5px solid #7fa89e;padding:8px 14px;border-radius:12px;border:1px solid #3a4a52;margin:18px 0 10px 0;">Why this uses conditional aggregation</h3>
<p style="margin:10px 0 15px 0;line-height:1.82;color:#d7e0e7;font-size:1rem;">We group by department, then inside each group we count how many rows satisfy a condition.</p>
</div>


In [7]:
high_performers_by_dept = df.groupby("dept").agg(
    high_performers=("rating", lambda x: (x >= 4).sum())
).reset_index()

high_performers_by_dept

,dept,high_performers
0,Data,3
1,ML,3
2,Web,1


<div data-notebook-cell-style-version="2026-04-v2" style="color:#d7e0e7;line-height:1.8;font-family:ui-sans-serif,system-ui,-apple-system,BlinkMacSystemFont,'Segoe UI',sans-serif;">
<h3 style="display:inline-block;color:#d8e4ea;background:#22303a;border-left:5px solid #7fa89e;padding:8px 14px;border-radius:12px;border:1px solid #3a4a52;margin:18px 0 10px 0;">Explanation</h3>
<p style="margin:10px 0 15px 0;line-height:1.82;color:#d7e0e7;font-size:1rem;">Inside each department:</p>
<ul style="margin:10px 0 16px 20px;padding-left:18px;line-height:1.9;color:#d2dce4;">
<li style="margin:6px 0;"><code style="background:#27313a;color:#e8eef4;padding:2px 6px;border-radius:6px;border:1px solid #3c4a55;">(x &gt;= 4)</code> creates booleans</li>
<li style="margin:6px 0;"><code style="background:#27313a;color:#e8eef4;padding:2px 6px;border-radius:6px;border:1px solid #3c4a55;">True</code> counts as <code style="background:#27313a;color:#e8eef4;padding:2px 6px;border-radius:6px;border:1px solid #3c4a55;">1</code></li>
<li style="margin:6px 0;"><code style="background:#27313a;color:#e8eef4;padding:2px 6px;border-radius:6px;border:1px solid #3c4a55;">False</code> counts as <code style="background:#27313a;color:#e8eef4;padding:2px 6px;border-radius:6px;border:1px solid #3c4a55;">0</code></li>
<li style="margin:6px 0;"><code style="background:#27313a;color:#e8eef4;padding:2px 6px;border-radius:6px;border:1px solid #3c4a55;">.sum()</code> gives total high performers</li>
</ul>
<p style="margin:10px 0 15px 0;line-height:1.82;color:#d7e0e7;font-size:1rem;">Example for <code style="background:#27313a;color:#e8eef4;padding:2px 6px;border-radius:6px;border:1px solid #3c4a55;">ML</code>, ratings are:</p>
<ul style="margin:10px 0 16px 20px;padding-left:18px;line-height:1.9;color:#d2dce4;">
<li style="margin:6px 0;">4, 5, 5</li>
</ul>
<p style="margin:10px 0 15px 0;line-height:1.82;color:#d7e0e7;font-size:1rem;">Condition <code style="background:#27313a;color:#e8eef4;padding:2px 6px;border-radius:6px;border:1px solid #3c4a55;">&gt;= 4</code> becomes:</p>
<ul style="margin:10px 0 16px 20px;padding-left:18px;line-height:1.9;color:#d2dce4;">
<li style="margin:6px 0;">True, True, True</li>
</ul>
<p style="margin:10px 0 15px 0;line-height:1.82;color:#d7e0e7;font-size:1rem;">Sum = <code style="background:#27313a;color:#e8eef4;padding:2px 6px;border-radius:6px;border:1px solid #3c4a55;">3</code></p>
</div>


<div data-notebook-cell-style-version="2026-04-v2" style="color:#d7e0e7;line-height:1.8;font-family:ui-sans-serif,system-ui,-apple-system,BlinkMacSystemFont,'Segoe UI',sans-serif;">
<h2 style="background:linear-gradient(90deg,#22313d 0%,#273743 100%);border-left:8px solid #72aa9b;color:#e7eff3;padding:11px 16px;border-radius:14px;border:1px solid #3a4b58;box-shadow:0 6px 18px rgba(0,0,0,0.18);margin:22px 0 14px 0;">7. Task 6 — Percentage of high performers by department</h2>
</div>


<div data-notebook-cell-style-version="2026-04-v2" style="color:#d7e0e7;line-height:1.8;font-family:ui-sans-serif,system-ui,-apple-system,BlinkMacSystemFont,'Segoe UI',sans-serif;">
<h3 style="display:inline-block;color:#d8e4ea;background:#22303a;border-left:5px solid #7fa89e;padding:8px 14px;border-radius:12px;border:1px solid #3a4a52;margin:18px 0 10px 0;">Goal</h3>
<p style="margin:10px 0 15px 0;line-height:1.82;color:#d7e0e7;font-size:1rem;">For each department, compute:</p>
<ul style="margin:10px 0 16px 20px;padding-left:18px;line-height:1.9;color:#d2dce4;">
<li style="margin:6px 0;">total employees</li>
<li style="margin:6px 0;">high performers</li>
<li style="margin:6px 0;">percentage of high performers</li>
</ul>
<h3 style="display:inline-block;color:#d8e4ea;background:#22303a;border-left:5px solid #7fa89e;padding:8px 14px;border-radius:12px;border:1px solid #3a4a52;margin:18px 0 10px 0;">Pattern</h3>
<p style="margin:10px 0 15px 0;line-height:1.82;color:#d7e0e7;font-size:1rem;">This is a two-step pattern:</p>
<ol style="margin:10px 0 16px 20px;padding-left:18px;line-height:1.9;color:#d2dce4;">
<li style="margin:6px 0;">aggregate counts</li>
<li style="margin:6px 0;">create a derived percentage column</li>
</ol>
</div>


In [8]:
high_perf_pct = df.groupby("dept").agg(
    total_employees=("emp_id", "count"),
    high_performers=("rating", lambda x: (x >= 4).sum())
).reset_index()

high_perf_pct["pct_high_performers"] = (
    high_perf_pct["high_performers"] / high_perf_pct["total_employees"] * 100
)

high_perf_pct

,dept,total_employees,high_performers,pct_high_performers
0,Data,4,3,75.000000
1,ML,3,3,100.000000
2,Web,3,1,33.333333


<div data-notebook-cell-style-version="2026-04-v2" style="color:#d7e0e7;line-height:1.8;font-family:ui-sans-serif,system-ui,-apple-system,BlinkMacSystemFont,'Segoe UI',sans-serif;">
<h3 style="display:inline-block;color:#d8e4ea;background:#22303a;border-left:5px solid #7fa89e;padding:8px 14px;border-radius:12px;border:1px solid #3a4a52;margin:18px 0 10px 0;">Explanation</h3>
<p style="margin:10px 0 15px 0;line-height:1.82;color:#d7e0e7;font-size:1rem;">For <code style="background:#27313a;color:#e8eef4;padding:2px 6px;border-radius:6px;border:1px solid #3c4a55;">Web</code>:</p>
<ul style="margin:10px 0 16px 20px;padding-left:18px;line-height:1.9;color:#d2dce4;">
<li style="margin:6px 0;">total employees = 3</li>
<li style="margin:6px 0;">high performers = 1</li>
<li style="margin:6px 0;">percentage = <code style="background:#27313a;color:#e8eef4;padding:2px 6px;border-radius:6px;border:1px solid #3c4a55;">1 / 3 * 100 = 33.33</code></li>
</ul>
<p style="margin:10px 0 15px 0;line-height:1.82;color:#d7e0e7;font-size:1rem;">This kind of question is common in interviews:</p>
<ul style="margin:10px 0 16px 20px;padding-left:18px;line-height:1.9;color:#d2dce4;">
<li style="margin:6px 0;">% premium users by country</li>
<li style="margin:6px 0;">% passed students by class</li>
<li style="margin:6px 0;">% fraud cases by region</li>
</ul>
</div>


<div data-notebook-cell-style-version="2026-04-v2" style="color:#d7e0e7;line-height:1.8;font-family:ui-sans-serif,system-ui,-apple-system,BlinkMacSystemFont,'Segoe UI',sans-serif;">
<h2 style="background:linear-gradient(90deg,#22313d 0%,#273743 100%);border-left:8px solid #72aa9b;color:#e7eff3;padding:11px 16px;border-radius:14px;border:1px solid #3a4b58;box-shadow:0 6px 18px rgba(0,0,0,0.18);margin:22px 0 14px 0;">8. Task 7 — Top paid employee in each department</h2>
</div>


<div data-notebook-cell-style-version="2026-04-v2" style="color:#d7e0e7;line-height:1.8;font-family:ui-sans-serif,system-ui,-apple-system,BlinkMacSystemFont,'Segoe UI',sans-serif;">
<h3 style="display:inline-block;color:#d8e4ea;background:#22303a;border-left:5px solid #7fa89e;padding:8px 14px;border-radius:12px;border:1px solid #3a4a52;margin:18px 0 10px 0;">Goal</h3>
<p style="margin:10px 0 15px 0;line-height:1.82;color:#d7e0e7;font-size:1rem;">Return the full row of the highest salary employee per department.</p>
<p style="margin:10px 0 15px 0;line-height:1.82;color:#d7e0e7;font-size:1rem;">There are two good interview-friendly ways to solve this:</p>
<ol style="margin:10px 0 16px 20px;padding-left:18px;line-height:1.9;color:#d2dce4;">
<li style="margin:6px 0;"><code style="background:#27313a;color:#e8eef4;padding:2px 6px;border-radius:6px;border:1px solid #3c4a55;">groupby + nlargest</code></li>
<li style="margin:6px 0;"><code style="background:#27313a;color:#e8eef4;padding:2px 6px;border-radius:6px;border:1px solid #3c4a55;">sort_values + drop_duplicates</code></li>
</ol>
</div>


<div data-notebook-cell-style-version="2026-04-v2" style="color:#d7e0e7;line-height:1.8;font-family:ui-sans-serif,system-ui,-apple-system,BlinkMacSystemFont,'Segoe UI',sans-serif;">
<h3 style="display:inline-block;color:#d8e4ea;background:#22303a;border-left:5px solid #7fa89e;padding:8px 14px;border-radius:12px;border:1px solid #3a4a52;margin:18px 0 10px 0;">Method 1 — <code style="background:#27313a;color:#e8eef4;padding:2px 6px;border-radius:6px;border:1px solid #3c4a55;">groupby + nlargest</code></h3>
</div>


In [9]:
top_paid_by_dept = df.groupby("dept", group_keys=False).apply(
    lambda g: g.nlargest(1, "salary")
)

top_paid_by_dept[["emp_id", "name", "dept", "salary"]]

/tmp/ipykernel_2824/390576734.py:1: DeprecationWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.
  top_paid_by_dept = df.groupby("dept", group_keys=False).apply(


,emp_id,name,dept,salary
4,105,Eva,Data,130000
9,110,Jane,ML,155000
2,103,Charlie,Web,110000


<div data-notebook-cell-style-version="2026-04-v2" style="color:#d7e0e7;line-height:1.8;font-family:ui-sans-serif,system-ui,-apple-system,BlinkMacSystemFont,'Segoe UI',sans-serif;">
<h3 style="display:inline-block;color:#d8e4ea;background:#22303a;border-left:5px solid #7fa89e;padding:8px 14px;border-radius:12px;border:1px solid #3a4a52;margin:18px 0 10px 0;">Explanation</h3>
<p style="margin:10px 0 15px 0;line-height:1.82;color:#d7e0e7;font-size:1rem;">Within each department group:</p>
<ul style="margin:10px 0 16px 20px;padding-left:18px;line-height:1.9;color:#d2dce4;">
<li style="margin:6px 0;">sort by salary descending</li>
<li style="margin:6px 0;">keep the largest 1 row</li>
</ul>
<p style="margin:10px 0 15px 0;line-height:1.82;color:#d7e0e7;font-size:1rem;">This returns the full employee row, not just the salary.</p>
</div>


<div data-notebook-cell-style-version="2026-04-v2" style="color:#d7e0e7;line-height:1.8;font-family:ui-sans-serif,system-ui,-apple-system,BlinkMacSystemFont,'Segoe UI',sans-serif;">
<h3 style="display:inline-block;color:#d8e4ea;background:#22303a;border-left:5px solid #7fa89e;padding:8px 14px;border-radius:12px;border:1px solid #3a4a52;margin:18px 0 10px 0;">Method 2 — <code style="background:#27313a;color:#e8eef4;padding:2px 6px;border-radius:6px;border:1px solid #3c4a55;">sort_values + drop_duplicates</code></h3>
</div>


In [10]:
top_paid_by_dept_2 = (
    df.sort_values(["dept", "salary"], ascending=[True, False])
      .drop_duplicates("dept")
)

top_paid_by_dept_2[["emp_id", "name", "dept", "salary"]]

,emp_id,name,dept,salary
4,105,Eva,Data,130000
9,110,Jane,ML,155000
2,103,Charlie,Web,110000


<div data-notebook-cell-style-version="2026-04-v2" style="color:#d7e0e7;line-height:1.8;font-family:ui-sans-serif,system-ui,-apple-system,BlinkMacSystemFont,'Segoe UI',sans-serif;">
<h3 style="display:inline-block;color:#d8e4ea;background:#22303a;border-left:5px solid #7fa89e;padding:8px 14px;border-radius:12px;border:1px solid #3a4a52;margin:18px 0 10px 0;">Explanation</h3>
<p style="margin:10px 0 15px 0;line-height:1.82;color:#d7e0e7;font-size:1rem;">This works because:</p>
<ol style="margin:10px 0 16px 20px;padding-left:18px;line-height:1.9;color:#d2dce4;">
<li style="margin:6px 0;">sort each department so the highest salary comes first</li>
<li style="margin:6px 0;">drop duplicates by department</li>
<li style="margin:6px 0;">first row kept per department is the top paid one</li>
</ol>
<p style="margin:10px 0 15px 0;line-height:1.82;color:#d7e0e7;font-size:1rem;">This method is also very common in interviews.</p>
</div>


<div data-notebook-cell-style-version="2026-04-v2" style="color:#d7e0e7;line-height:1.8;font-family:ui-sans-serif,system-ui,-apple-system,BlinkMacSystemFont,'Segoe UI',sans-serif;">
<h2 style="background:linear-gradient(90deg,#22313d 0%,#273743 100%);border-left:8px solid #72aa9b;color:#e7eff3;padding:11px 16px;border-radius:14px;border:1px solid #3a4b58;box-shadow:0 6px 18px rgba(0,0,0,0.18);margin:22px 0 14px 0;">Final cheat sheet summary</h2>
<h3 style="display:inline-block;color:#d8e4ea;background:#22303a;border-left:5px solid #7fa89e;padding:8px 14px;border-radius:12px;border:1px solid #3a4a52;margin:18px 0 10px 0;">Use these patterns clearly</h3>
<h4 style="color:#9fd2c3;margin:16px 0 8px 0;font-weight:700;">1. Summary per group</h4>
<pre style="background:#1b2330;color:#e6edf3;padding:14px 16px;border-radius:14px;border:1px solid #334155;overflow-x:auto;line-height:1.7;margin:14px 0;"><code class="language-python" style="background:transparent;color:#e6edf3;border:0;padding:0;">df.groupby(&quot;dept&quot;).agg(total_salary=(&quot;salary&quot;, &quot;sum&quot;))
</code></pre>
<h4 style="color:#9fd2c3;margin:16px 0 8px 0;font-weight:700;">2. Summary by multiple keys</h4>
<pre style="background:#1b2330;color:#e6edf3;padding:14px 16px;border-radius:14px;border:1px solid #334155;overflow-x:auto;line-height:1.7;margin:14px 0;"><code class="language-python" style="background:transparent;color:#e6edf3;border:0;padding:0;">df.groupby([&quot;dept&quot;, &quot;city&quot;], as_index=False).agg(avg_salary=(&quot;salary&quot;, &quot;mean&quot;))
</code></pre>
<h4 style="color:#9fd2c3;margin:16px 0 8px 0;font-weight:700;">3. Group metric back to each row</h4>
<pre style="background:#1b2330;color:#e6edf3;padding:14px 16px;border-radius:14px;border:1px solid #334155;overflow-x:auto;line-height:1.7;margin:14px 0;"><code class="language-python" style="background:transparent;color:#e6edf3;border:0;padding:0;">df[&quot;dept_avg_salary&quot;] = df.groupby(&quot;dept&quot;)[&quot;salary&quot;].transform(&quot;mean&quot;)
</code></pre>
<h4 style="color:#9fd2c3;margin:16px 0 8px 0;font-weight:700;">4. Condition using a group metric</h4>
<pre style="background:#1b2330;color:#e6edf3;padding:14px 16px;border-radius:14px;border:1px solid #334155;overflow-x:auto;line-height:1.7;margin:14px 0;"><code class="language-python" style="background:transparent;color:#e6edf3;border:0;padding:0;">df[&quot;flag&quot;] = np.where(df[&quot;salary&quot;] &gt; df[&quot;dept_avg_salary&quot;], &quot;above&quot;, &quot;below&quot;)
</code></pre>
<h4 style="color:#9fd2c3;margin:16px 0 8px 0;font-weight:700;">5. Conditional count inside groupby</h4>
<pre style="background:#1b2330;color:#e6edf3;padding:14px 16px;border-radius:14px;border:1px solid #334155;overflow-x:auto;line-height:1.7;margin:14px 0;"><code class="language-python" style="background:transparent;color:#e6edf3;border:0;padding:0;">df.groupby(&quot;dept&quot;).agg(high_performers=(&quot;rating&quot;, lambda x: (x &gt;= 4).sum()))
</code></pre>
<h4 style="color:#9fd2c3;margin:16px 0 8px 0;font-weight:700;">6. Top row per group</h4>
<pre style="background:#1b2330;color:#e6edf3;padding:14px 16px;border-radius:14px;border:1px solid #334155;overflow-x:auto;line-height:1.7;margin:14px 0;"><code class="language-python" style="background:transparent;color:#e6edf3;border:0;padding:0;">df.groupby(&quot;dept&quot;, group_keys=False).apply(lambda g: g.nlargest(1, &quot;salary&quot;))
</code></pre>
<h3 style="display:inline-block;color:#d8e4ea;background:#22303a;border-left:5px solid #7fa89e;padding:8px 14px;border-radius:12px;border:1px solid #3a4a52;margin:18px 0 10px 0;">Interview memory line</h3>
<ul style="margin:10px 0 16px 20px;padding-left:18px;line-height:1.9;color:#d2dce4;">
<li style="margin:6px 0;"><code style="background:#27313a;color:#e8eef4;padding:2px 6px;border-radius:6px;border:1px solid #3c4a55;">agg</code> = one row per group</li>
<li style="margin:6px 0;"><code style="background:#27313a;color:#e8eef4;padding:2px 6px;border-radius:6px;border:1px solid #3c4a55;">transform</code> = same row count</li>
<li style="margin:6px 0;"><code style="background:#27313a;color:#e8eef4;padding:2px 6px;border-radius:6px;border:1px solid #3c4a55;">apply</code> = custom group logic</li>
<li style="margin:6px 0;"><code style="background:#27313a;color:#e8eef4;padding:2px 6px;border-radius:6px;border:1px solid #3c4a55;">np.where</code> = if/else column creation</li>
</ul>
</div>
